### PR TITLE
Make computing snapshot checksums faster in debug

### PR DESCRIPTION
**Pull request by timvisee**
Computing file checksums is extremely slow in our debug builds. I was experimenting with snapshots again today and decided to fix this because it was quite annoying.

This enables the release optimizations in all profiles for the sha2 sub crate to always benefit from the faster implementation.

Before: checksum takes 15.6s
After: checksum takes 0.4s

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you formatted your code locally using `cargo +nightly fmt --all` command prior to submission?
3. [x] Have you checked your code using `cargo clippy --all --all-features` command?

### CODE REVIEW

**Comment by ffuugoo** [comment]
On macOS it's just `shasum`... 😂
Screenshot 2024-09-11 at 11 39 40

Though, I don't think we care about macOS much.

**Comment by timvisee** [comment]
> On macOS it's just `shasum`... 😂 Screenshot 2024-09-11 at 11 39 40
> 
> Though, I don't think we care about macOS much.

Oh noes!

Thanks for checking. I'll validate on my end how likely it is for `sha256sum` to exist. Changing the compile time conditions is probably good enough.

**Comment by timvisee** [comment]
I asked in our internal channels, and all Linux users do seem to have it.

Linux does also have `shasum` but it's about 2.5 times slower in my quick tests.

I'll just enable it on Linux only then to keep things simple. It's not very significant anyway and falls back if the binary is not there.

File: lib/collection/src/common/sha_256.rs [hunk 1]
```diff
@@ -25,6 +41,52 @@ pub async fn hash_file(file_path: &Path) -> io::Result<String> {
...
+/// Compute sha256 hash for the given file using the `sha256sum` binary
+///
+/// May fail if the binary does not exist on the system.
+#[cfg(all(debug_assertions, unix))]
+async fn hash_file_bin(file_path: &Path) -> io::Result<String> {
+    use tokio::process::Command;
+
+    let output = match Command::new("sha256sum")
+        .arg("-b")
+        .arg(file_path.as_os_str())
```
Comment (ffuugoo) [inline lib/collection/src/common/sha_256.rs:53]: I think just `file_path` would work here.

```suggestion
        .arg(file_path)
```

File: lib/collection/src/common/sha_256.rs [hunk 2]
```diff
@@ -25,6 +41,52 @@ pub async fn hash_file(file_path: &Path) -> io::Result<String> {
...
+            ));
+        }
+        // Binary must have been called successfully
+        Err(err) => {
+            return Err(io::Error::new(
+                io::ErrorKind::Other,
+                format!("Failed to compute checksum with sha256sum: {err}"),
+            ));
+        }
+        Ok(command) => command,
```
Comment (ffuugoo) [inline lib/collection/src/common/sha_256.rs:74]: I'd use `Ok(command) if command.status.success() => command` at the top.

```suggestion
        Ok(command) if command.status.success() => command,
        Ok(command) => {
            return Err(io::Error::new(
                io::ErrorKind::Other,
                format!(
                    "Failed to compute checksum with sha256sum: {}",
                    String::from_utf8(command.stderr).unwrap(),
                ),
            ));
        }
        Err(err) => {
            return Err(io::Error::new(
                io::ErrorKind::Other,
                format!("Failed to compute checksum with sha256sum: {err}"),
            ));
        }
```

**Review by ffuugoo — APPROVED** [review]
Feel free to ignore comments, cause they are bikesheddiest of bikesheddings.

**Review by generall — APPROVED** [review]
`sha256sum` exists even in our docker container (which is not even required) 
So I assume this this fine.

An alternative I would consider is to just disable checksum in dev

File: lib/collection/src/common/sha_256.rs [hunk 3]
```diff
@@ -25,6 +41,50 @@ pub async fn hash_file(file_path: &Path) -> io::Result<String> {
...
+/// Compute sha256 hash for the given file using the `sha256sum` binary
+///
+/// May fail if the binary does not exist on the system.
+#[cfg(all(debug_assertions, target_os = "linux"))]
+async fn hash_file_bin(file_path: &Path) -> io::Result<String> {
+    use tokio::process::Command;
+
+    let output = match Command::new("sha256sum")
+        .arg("-b")
+        .arg(file_path)
```
Comment (xzfc) [inline lib/collection/src/common/sha_256.rs:53]: Security/correctness: prepend `--` argument when passing arbitrary filenames so weird filenames like `--help` won't be treated as options.

```suggestion
    let output = match Command::new("sha256sum")
        .arg("-b")
        .arg("--")
        .arg(file_path)
```

**Comment by ffuugoo** [comment]
> We could go with [`Cargo.toml` overrides](https://doc.rust-lang.org/cargo/reference/profiles.html#overrides), i.e. by adding this to our top-level `Cargo.toml`:
> 
> ```toml
> [profile.dev.package.sha2]
> opt-level = 3
> ```

🤯

I vote to test this one first, then.

File: lib/collection/src/common/sha_256.rs (hunk 3 above)
Comment (timvisee) [reply lib/collection/src/common/sha_256.rs:53]: Thanks. Didn't think about that one!

**Comment by timvisee** [comment]
> > Ideally we enable release optimizations for the built-in hasher at all times. But I don't see a proper way to achieve this.
> 
> We could go with [`Cargo.toml` overrides](https://doc.rust-lang.org/cargo/reference/profiles.html#overrides), i.e. by adding this to our top-level `Cargo.toml`:
> 
> ```toml
> [profile.dev.package.sha2]
> opt-level = 3
> ```

Woah! I wasn't aware of this 🥹

Let me give it a quick shot. I'll definitely go with that if that works.

**Comment by timvisee** [comment]
Brilliant, works like a charm!

Latest test for hashing duration:
- before: 15.6s
- with `sha256sum`: 0.3s (got to love GNU coretutils)
- with release mode: 0.4s

That's obviously good enough. I've updated the PR to enable release optimizations for the sub-crate.

**Pull request merged**